scripts/cluster-upgrade.sh

```diff
@@ -0,0 +1,161 @@
+
+    GRAPH_EDGE=$(echo "${GRAPH}" | jq -r ".edges[] | select(.[0] == $GRAPH_INDEX_FROM) | .[1] == $GRAPH_INDEX_TO" | grep true)
+
+    if [ -z $GRAPH_INDEX_FROM ];
+    then
+        echo "Cannot upgrade from $OCP_VERSION_FROM, it is not available in upgrade graph."
+        exit 2
+    fi
+    if [ -z $GRAPH_INDEX_TO ];
+    then
+        echo "Cannot upgrade to $OCP_VERSION_TO, it is not available in upgrade graph."
+        exit 2
+    fi
+    if [ "$GRAPH_EDGE" != "true" ];
+    then
+        echo "Cannot upgrade from $OCP_VERSION_FROM to $OCP_VERSION_TO, no path exists in upgrade graph."
+        exit 2
+    fi
+
+    # Verify the target version exists as a ClusterImageSet (CIS) on the cluster
+    # (if not, the customer could never install to that target version)
+    CIS_VERSION=$(oc get clusterimageset --all-namespaces -o json | jq -r ".items[].spec.releaseImage | split(\":\")[1] | select(. == \"$OCP_VERSION_TO\")")
+    if [ "$OCP_VERSION_TO" != "$CIS_VERSION" ];
+    then
+        echo "Cannot upgrade to $OCP_VERSION_TO, it is not available as a ClusterImageSet in the cluster."
+        exit 3
+    fi
+
+    # Verify we have ClusterDeployment CRs to work with
+    if [ `oc get crd clusterdeployments.hive.openshift.io --no-headers 2>/dev/null | wc -l` == "0" ];
+    then
+        echo "ERROR: Current cluster does not have Hive installed.  Verify where you are logged in."
+        exit 1
+    fi
+fi
+
+TMP_DIR=`mktemp -d`
+OCP_PATCH="{\"spec\":{\"desiredUpdate\": {\"force\": false, \"image\": \"\", \"version\": \"$OCP_VERSION_TO\"}}}"
+if [[ $? -ne 0 ]];
+then
+    echo "ERROR: Couldn't create a temporary directory. Aborting.
+    exit 1
+fi
+# kick off the upgrades (or status check)
+for CD_NAMESPACE in `oc get clusterdeployment --all-namespaces | awk '{print $1}' | sort | uniq`;
+do  
+    for CD_NAME in `oc -n $CD_NAMESPACE get clusterdeployment -o json | jq -r '.items[] | select(.metadata.labels["api.openshift.com/managed"] == "true") | select(.status.installed == true) | select(.status.clusterVersionStatus.history[0].state == "Completed") | .metadata.name'`;
+    do  
+        echo -n "Checking $CD_NAME..."
+        # TODO figure out how to get the latest version, for now just getting the first one (ordered descending)
+        OCP_CURRENT_VERSION=`oc -n $CD_NAMESPACE get clusterdeployment $CD_NAME -o json | jq -r '.status.clusterVersionStatus.history[] | select(.state == "Completed") | .version' | head -n1`
+
+        SS_NAME=$CD_NAME-upgrade
+        SS_VERSION_FROM=`oc -n $CD_NAMESPACE get syncset $SS_NAME --show-labels --no-headers 2> /dev/null | sed 's|.*managed.openshift.io/from=\([^,]*\).*|\1|g'`
+        SS_VERSION_TO=`oc -n $CD_NAMESPACE get syncset $SS_NAME --show-labels --no-headers 2> /dev/null | sed 's|.*managed.openshift.io/to=\([^,]*\).*|\1|g'`
+        SS_CREATED_ON=$(date --date="$(oc -n $CD_NAMESPACE get syncset $SS_NAME --template '{{ .metadata.creationTimestamp }}' 2>/dev/null)" +"%s" 2>/dev/null)
```

Comment:
Note that `date --date` is not portable. I don't know if it will come up, but in testing, if this is run on MacOS it could be an issue.

Reply:
You have to run this on the hive cluster to use it, so it's good enough in my view.  You could hit hive from MacOS if you tunnel through bastion I guess, but it's not the expected usage.

Reply:
Sounds good. In that case can I ask for a short comment at the top of the script saying that it's Linux only by design, to run on Hive cluster? I don't think we have to go through capabilities testing like we're autoconf. (I can make the suggested change for review)